File: scratch/t2-p2-worker-10.js
const fs = require('fs');
const path = require('path');
const { VertexAI } = require('@google-cloud/vertexai');

// Need to set Google app credentials
process.env.GOOGLE_APPLICATION_CREDENTIALS = path.resolve(process.cwd(), 'google-service-account.json');

// Get project ID from service account
const sa = JSON.parse(fs.readFileSync(process.env.GOOGLE_APPLICATION_CREDENTIALS, 'utf8'));
const projectId = sa.project_id;
const location = 'us-central1';

const vertex_ai = new VertexAI({project: projectId, location: location});
const model = vertex_ai.getGenerativeModel({
  model: 'gemini-2.5-flash-lite',
  generationConfig: { responseMimeType: 'application/json' }
});

async function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function translateBatch(batch, batchIndex) {
  console.log(`[Batch ${batchIndex}] Translating ${batch.length} items...`);
  const prompt = `You are a professional translator. I will provide a JSON array of objects.
Each object has a "taskId", "type", "loc" (target locale code), and "originalEn" (the text to translate, which may be Korean or English).
- If "type" is 'fact-layer', "originalEn" is an array of { year, event } objects. Translate BOTH year and event into the target "loc". Keep the format identical.
- If "type" is 'narrative', "originalEn" is a string. Translate it into the target "loc".

Return a JSON array of objects. Each object MUST have:
- "taskId": the same taskId as provided
- "translated": the translated content (either array of {year, event} or string).

INPUT JSON ARRAY:
${JSON.stringify(batch.map((b, i) => ({
  taskId: i,
  loc: b.loc,
  type: b.type,
  originalEn: b.originalEn
})), null, 2)}
`;

  let attempts = 0;
  while (attempts < 3) {
    try {
      const result = await model.generateContent(prompt);
      const text = result.response.candidates[0].content.parts[0].text;
      let parsed;
      try {
        parsed = JSON.parse(text);
      } catch (e) {
        const match = text.match(/\[[\s\S]*\]/);
        if (match) {
          parsed = JSON.parse(match[0]);
        } else {
          throw new Error("No JSON array found in response");
        }
      }
      return parsed;
    } catch (err) {
      attempts++;
      console.error(`[Batch ${batchIndex}] Attempt ${attempts} failed: ${err.message}`);
      await sleep(2000 * attempts);
    }
  }
  throw new Error(`Failed to translate batch ${batchIndex}`);
}

async function main() {
  const inputPath = path.resolve('scratch/t2-p2-chunk-10.json');
  const outputPath = path.resolve('scratch/t2-p2-out-10.json');

  const rawData = fs.readFileSync(inputPath, 'utf8');
  const data = JSON.parse(rawData);

  console.log(`Total items to translate: ${data.length}`);

  const batchSize = 10;
  const outData = [];

  for (let i = 0; i < data.length; i += batchSize) {
    const batch = data.slice(i, i + batchSize);
    const translatedBatchRaw = await translateBatch(batch, Math.floor(i / batchSize));
    
    for (let j = 0; j < batch.length; j++) {
      const item = batch[j];
      const translatedItemRaw = translatedBatchRaw.find(t => t.taskId === j);
      outData.push({
        ...item,
        translated: translatedItemRaw ? translatedItemRaw.translated : null
      });
    }
    
    // Save intermediate progress just in case
    fs.writeFileSync(outputPath, JSON.stringify(outData, null, 2), 'utf8');
    await sleep(500); // rate limiting
  }

  console.log(`Finished translating. Saved to ${outputPath}`);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scratch/t2-p2-worker-10.py
import json
import os
import time
import sys
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')
from deep_translator import GoogleTranslator

def translate_item(item, max_retries=3):
    loc = item['loc']
    target_loc = loc
    # map any specific loc codes if deep_translator requires it
    # 'id' is standard, 'vi' is standard, etc.
    
    translator = GoogleTranslator(source='auto', target=target_loc)
    
    typ = item['type']
    orig = item['originalEn']
    
    for attempt in range(max_retries):
        try:
            if typ == 'fact-layer':
                translated = []
                for ev in orig:
                    if len(ev['year'].strip()) == 0:
                        y_trans = ev['year']
                    elif ev['year'].isdigit():
                        y_trans = ev['year']
                    else:
                        y_trans = translator.translate(ev['year'])
                    
                    e_trans = translator.translate(ev['event']) if ev['event'].strip() else ev['event']
                    translated.append({'year': y_trans, 'event': e_trans})
                
                return translated
            else:
                # narrative is a string, which might be long.
                # deep_translator limit is 5000 chars.
                if len(orig) <= 4900:
                    return translator.translate(orig)
                else:
                    # split by paragraph or sentences if needed.
                    # A simple approach: split by double newline.
                    parts = orig.split('\n\n')
                    translated_parts = []
                    for p in parts:
                        if not p.strip():
                            translated_parts.append(p)
                        else:
                            # if a single paragraph is > 4900 chars, split by sentences.
                            if len(p) > 4900:
                                subparts = [p[i:i+4900] for i in range(0, len(p), 4900)]
                                translated_parts.extend([translator.translate(sp) for sp in subparts])
                            else:
                                translated_parts.append(translator.translate(p))
                    return '\n\n'.join(translated_parts)
                    
        except Exception as e:
            print(f"Error on attempt {attempt+1}: {e}")
            time.sleep(2)
            
    print(f"Failed to translate task {item.get('taskId', 'unknown')}")
    return orig # fallback

def main():
    input_path = "scratch/t2-p2-chunk-10.json"
    output_path = "scratch/t2-p2-out-10.json"
    
    with open(input_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
        
    print(f"Total items: {len(data)}")
    
    out_data = []
    
    for i, item in enumerate(data):
        print(f"Translating item {i+1}/{len(data)} [loc: {item['loc']}, type: {item['type']}]")
        translated_content = translate_item(item)
        
        new_item = item.copy()
        new_item['translated'] = translated_content
        out_data.append(new_item)
        
        # Save every 20 items to not lose progress
        if (i + 1) % 20 == 0:
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(out_data, f, ensure_ascii=False, indent=2)
            time.sleep(1) # throttle a bit to avoid ban
            
    # Final save
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(out_data, f, ensure_ascii=False, indent=2)
        
    print(f"Successfully finished translating all items. Saved to {output_path}")

if __name__ == "__main__":
    main()
